lib/Sema/MiscDiagnostics.cpp

```diff
@@ -2639,13 +2687,29 @@ static void checkCStyleForLoop(TypeChecker &TC, const ForStmt *FS) {
   SourceLoc endOfIncrementLoc =
     Lexer::getLocForEndOfToken(TC.Context.SourceMgr,
                                FS->getIncrement().getPtrOrNull()->getEndLoc());
-    
-  diagnostic
-   .fixItRemoveChars(loopVarDecl->getLoc(), loopVar->getLoc())
-   .fixItReplaceChars(loopPatternEnd, startValue->getStartLoc(), " in ")
-   .fixItReplaceChars(FS->getFirstSemicolonLoc(), endValue->getStartLoc(),
-                      " ..< ")
-   .fixItRemoveChars(FS->getSecondSemicolonLoc(), endOfIncrementLoc);
+
+  if (strideByOne && OpKind != OperatorKind::Greater) {
+    diagnostic
+    .fixItRemoveChars(loopVarDecl->getLoc(), loopVar->getLoc())
+    .fixItReplaceChars(loopPatternEnd, startValue->getStartLoc(), " in ")
+    .fixItReplaceChars(FS->getFirstSemicolonLoc(), endValue->getStartLoc(),
+                       " ..< ")
+    .fixItRemoveChars(FS->getSecondSemicolonLoc(), endOfIncrementLoc);
+    return;
+  } else if (strideBackByOne && OpKind != OperatorKind::Smaller) {
+    SourceLoc startValueEnd = Lexer::getLocForEndOfToken(TC.Context.SourceMgr,
+                                                         startValue->getEndLoc());
+
+    StringRef endValueStr = CharSourceRange(TC.Context.SourceMgr, endValue->getStartLoc(),
+      Lexer::getLocForEndOfToken(TC.Context.SourceMgr, endValue->getEndLoc())).str();
+
+    diagnostic
+    .fixItRemoveChars(loopVarDecl->getLoc(), loopVar->getLoc())
+    .fixItReplaceChars(loopPatternEnd, startValue->getStartLoc(), " in ")
+    .fixItInsert(startValue->getStartLoc(), (llvm::Twine("((") + endValueStr + " + 1)...").str())
```

Comment:
Thanks for the comments! I am addressing the rest of them. The worst case here will be an error on creating ranges by non-integer values, I believe users will notice that. Thoughts?